### DISCUSSION TITLE
Watch the size of /boot when running updates

### DISCUSSION THREAD

**Discussion by RagnarSTS** [Security-Onion-Solutions/securityonion#11859] [Ideas] [answered]
I just had a sensor node that had a /boot of 500M that was too full to do upgrades.
IIRC the system started on 2.4.10, upgraded to 2.4.20, and failed temporarily bumping to 2.4.30.  
I think it had a pair of 5.14 and 5.15 kernels and lacked a bit of space for a 5th.  Just need to keep things tidy in the /boot.

**Comment by dougburks** [comment]
Was this sensor installed from our Security Onion ISO image or did you perform a network installation?

**Comment by RagnarSTS** [comment]
It was seco 2.4.10-20230821.iso

**Reply by dougburks** [reply]
If you're using our Security Onion ISO image (based on Oracle Linux), then we should automatically set `installonly_limit=2`:
https://github.com/Security-Onion-Solutions/securityonion/blob/2.4/main/salt/repo/client/files/oracle/yum.conf.jinja#L11

The `installonly_limit` setting is supposed to limit the number of kernels that remain on disk:
https://blogs.oracle.com/scoter/post/manage-the-number-of-installed-kernels-for-oracle-linux

Can you confirm that the `installonly_limit` setting is present on the sensor node in question?
```
sudo grep installonly_limit /etc/yum.conf
```

Can you also provide the full output of the following commands on the sensor node in question?
```
rpm -qa |grep kernel
```

```
sudo dnf update
```

```
ls -alh /boot
```

**Reply by RagnarSTS** [reply]
Well I manually cleaned up the original node.  
Installonly_limit is indeed set at 2 on all nodes.
```
rpm -qa | grep kernel:
kernel-uek-core-5.15.0-105.125.6.2.2.el9uek.x86_64
kernel-uek-modules-5.15.0-105.125.6.2.2.el9uek.x86_64
kernel-uek-5.15.0-105.125.6.2.2.el9uek.x86_64
kernel-tools-libs-5.14.0-362.8.1.el9_3.x86_64
kernel-modules-core-5.14.0-362.8.1.el9_3.x86_64
kernel-core-5.14.0-362.8.1.el9_3.x86_64
kernel-modules-5.14.0-362.8.1.el9_3.x86_64
kernel-uek-core-5.15.0-200.131.27.el9uek.x86_64
kernel-uek-modules-5.15.0-200.131.27.el9uek.x86_64
kernel-uek-5.15.0-200.131.27.el9uek.x86_64
kernel-5.14.0-362.8.1.el9_3.x86_64
kernel-tools-5.14.0-362.8.1.el9_3.x86_64
```

So maybe it isn't counting 'total' ... only either the oracle unbreakable kernel or standard kernels?

I manually cleaned out the troubled node:
```
ls -alh /boot
rw-r--r--.  1 root root 236K Sep 20 06:33 config-5.15.0-105.125.6.2.2.el9uek.x86_64
-rw-r--r--.  1 root root 237K Oct  5 05:14 config-5.15.0-200.131.27.el9uek.x86_64
drwxr-xr-x.  3 root root 4.0K Sep  2 16:33 efi
drwx------.  5 root root 4.0K Nov 22 05:02 grub2
-rw-------.  1 root root  87M Sep  2 16:41 initramfs-0-rescue-c5501468efca49f3a666186cfb6ce7a0.img
-rw-------.  1 root root  34M Sep 24 10:28 initramfs-5.15.0-105.125.6.2.2.el9uek.x86_64.img
-rw-------.  1 root root  32M Nov 22 05:04 initramfs-5.15.0-105.125.6.2.2.el9uek.x86_64kdump.img
-rw-------.  1 root root  34M Nov 22 05:04 initramfs-5.15.0-200.131.27.el9uek.x86_64.img
-rw-------.  1 root root  32M Nov 22 05:20 initramfs-5.15.0-200.131.27.el9uek.x86_64kdump.img
drwxr-xr-x.  3 root root 4.0K Sep  2 16:34 loader
drwx------.  2 root root  16K Sep  2 16:32 lost+found
lrwxrwxrwx.  1 root root   58 Sep 24 10:27 symvers-5.15.0-105.125.6.2.2.el9uek.x86_64.gz -> /lib/modules/5.15.0-105.125.6.2.2.el9uek.x86_64/symvers.gz
lrwxrwxrwx.  1 root root   55 Nov 22 05:04 symvers-5.15.0-200.131.27.el9uek.x86_64.gz -> /lib/modules/5.15.0-200.131.27.el9uek.x86_64/symvers.gz
-rw-------.  1 root root 6.3M Sep 20 06:33 System.map-5.15.0-105.125.6.2.2.el9uek.x86_64
-rw-------.  1 root root 6.3M Oct  5 05:14 System.map-5.15.0-200.131.27.el9uek.x86_64
-rwxr-xr-x.  1 root root  13M Sep  2 16:40 vmlinuz-0-rescue-c5501468efca49f3a666186cfb6ce7a0
-rw-r--r--.  1 root root  166 Nov 13 22:11 .vmlinuz-5.14.0-362.8.1.el9_3.x86_64.hmac
-rwxr-xr-x.  1 root root  13M Sep 20 06:33 vmlinuz-5.15.0-105.125.6.2.2.el9uek.x86_64
-rw-r--r--.  1 root root  179 Sep 20 06:32 .vmlinuz-5.15.0-105.125.6.2.2.el9uek.x86_64.hmac
-rwxr-xr-x.  1 root root  13M Oct  5 05:14 vmlinuz-5.15.0-200.131.27.el9uek.x86_64
-rw-r--r--.  1 root root  176 Oct  5 05:12 .vmlinuz-5.15.0-200.131.27.el9uek.x86_64.hmac
```

Compare to mastersearch node that was not manually beat into submission:

```
rw-r--r--.  1 root root 211K Sep 14 18:44 config-5.14.0-284.30.0.1.el9_2.x86_64
-rw-r--r--.  1 root root 215K Nov 13 22:12 config-5.14.0-362.8.1.el9_3.x86_64
-rw-r--r--.  1 root root 236K Sep 20 06:33 config-5.15.0-105.125.6.2.2.el9uek.x86_64
-rw-r--r--.  1 root root 237K Oct  5 05:14 config-5.15.0-200.131.27.el9uek.x86_64
drwxr-xr-x.  3 root root   17 Sep  2 16:02 efi
drwx------.  5 root root   97 Nov 22 03:03 grub2
-rw-------.  1 root root  87M Sep  2 16:07 initramfs-0-rescue-f270d7b55f5647d38f799dbc978a7bf5.img
-rw-------.  1 root root  33M Sep 17 07:02 initramfs-5.14.0-284.30.0.1.el9_2.x86_64.img
-rw-------.  1 root root  34M Nov 18 06:10 initramfs-5.14.0-362.8.1.el9_3.x86_64.img
-rw-------.  1 root root  33M Sep 24 09:55 initramfs-5.15.0-105.125.6.2.2.el9uek.x86_64.img
-rw-------.  1 root root  32M Nov 20 07:01 initramfs-5.15.0-105.125.6.2.2.el9uek.x86_64kdump.img
-rw-------.  1 root root  34M Nov 18 06:11 initramfs-5.15.0-200.131.27.el9uek.x86_64.img
-rw-------.  1 root root  32M Nov 21 02:02 initramfs-5.15.0-200.131.27.el9uek.x86_64kdump.img
drwxr-xr-x.  3 root root   21 Sep  2 16:03 loader
lrwxrwxrwx.  1 root root   54 Sep 17 07:01 symvers-5.14.0-284.30.0.1.el9_2.x86_64.gz -> /lib/modules/5.14.0-284.30.0.1.el9_2.x86_64/symvers.gz
lrwxrwxrwx.  1 root root   51 Nov 18 06:10 symvers-5.14.0-362.8.1.el9_3.x86_64.gz -> /lib/modules/5.14.0-362.8.1.el9_3.x86_64/symvers.gz
lrwxrwxrwx.  1 root root   58 Sep 24 09:54 symvers-5.15.0-105.125.6.2.2.el9uek.x86_64.gz -> /lib/modules/5.15.0-105.125.6.2.2.el9uek.x86_64/symvers.gz
lrwxrwxrwx.  1 root root   55 Nov 18 06:10 symvers-5.15.0-200.131.27.el9uek.x86_64.gz -> /lib/modules/5.15.0-200.131.27.el9uek.x86_64/symvers.gz
-rw-------.  1 root root 5.7M Sep 14 18:44 System.map-5.14.0-284.30.0.1.el9_2.x86_64
-rw-------.  1 root root 8.1M Nov 13 22:12 System.map-5.14.0-362.8.1.el9_3.x86_64
-rw-------.  1 root root 6.3M Sep 20 06:33 System.map-5.15.0-105.125.6.2.2.el9uek.x86_64
-rw-------.  1 root root 6.3M Oct  5 05:14 System.map-5.15.0-200.131.27.el9uek.x86_64
-rwxr-xr-x.  1 root root  13M Sep  2 16:07 vmlinuz-0-rescue-f270d7b55f5647d38f799dbc978a7bf5
-rwxr-xr-x.  1 root root  12M Sep 14 18:44 vmlinuz-5.14.0-284.30.0.1.el9_2.x86_64
-rw-r--r--.  1 root root  169 Sep 14 18:43 .vmlinuz-5.14.0-284.30.0.1.el9_2.x86_64.hmac
-rwxr-xr-x.  1 root root  13M Nov 13 22:12 vmlinuz-5.14.0-362.8.1.el9_3.x86_64
-rw-r--r--.  1 root root  166 Nov 13 22:11 .vmlinuz-5.14.0-362.8.1.el9_3.x86_64.hmac
-rwxr-xr-x.  1 root root  13M Sep 20 06:33 vmlinuz-5.15.0-105.125.6.2.2.el9uek.x86_64
-rw-r--r--.  1 root root  179 Sep 20 06:32 .vmlinuz-5.15.0-105.125.6.2.2.el9uek.x86_64.hmac
-rwxr-xr-x.  1 root root  13M Oct  5 05:14 vmlinuz-5.15.0-200.131.27.el9uek.x86_64
-rw-r--r--.  1 root root  176 Oct  5 05:12 .vmlinuz-5.15.0-200.131.27.el9uek.x86_64.hmac
```

dnf update noted nothing to do anywhare.

**Reply by RagnarSTS** [reply]
Of note the sensor I manually removed things from shows 62% use on /boot with 173M available.  
The mastersearch that wasn't manually poked at shows 82% usage of /boot with 96M available.

**Reply by dougburks** [reply]
I tried to duplicate your experience but was unable to. I started with our 2.4.4 RC1 ISO image and installed in airgap mode so that each run of soup would install OS updates as well. I then soup'd to 2.4.5, 2.4.10, 2.4.20, and finally 2.4.30 but never filled /boot. Old kernels were automatically purged as expected. Further details below. Did you install any extra packages or store any extra data in /boot?

After installing 2.4.4 RC1:
```
[root@securityonion ~]# cat /etc/soversion 
2.4.4
[root@securityonion ~]# rpm -qa |egrep "kernel-5|kernel-uek-5"
kernel-5.14.0-284.18.1.el9_2.x86_64
kernel-uek-5.15.0-103.114.4.el9uek.x86_64
[root@securityonion ~]# df -h /boot
Filesystem      Size  Used Avail Use% Mounted on
/dev/sda2       507M  278M  230M  55% /boot
```

After soup to 2.4.5 RC2:
```
[root@securityonion ~]# cat /etc/soversion 
2.4.5
[root@securityonion ~]# rpm -qa |egrep "kernel-5|kernel-uek-5"
kernel-5.14.0-284.18.1.el9_2.x86_64
kernel-uek-5.15.0-103.114.4.el9uek.x86_64
kernel-5.14.0-284.25.1.el9_2.x86_64
[root@securityonion ~]# df -h /boot
Filesystem      Size  Used Avail Use% Mounted on
/dev/sda2       507M  328M  179M  65% /boot
```

After soup to 2.4.10 GA:
```
[root@securityonion ~]# cat /etc/soversion 
2.4.10
[root@securityonion ~]# rpm -qa |egrep "kernel-5|kernel-uek-5"
kernel-uek-5.15.0-103.114.4.el9uek.x86_64
kernel-5.14.0-284.25.1.el9_2.x86_64
kernel-5.14.0-284.25.1.0.1.el9_2.x86_64
[root@securityonion ~]# df -h /boot
Filesystem      Size  Used Avail Use% Mounted on
/dev/sda2       507M  324M  183M  64% /boot
```

After soup to 2.4.20:
```
[root@securityonion ~]# cat /etc/soversion 
2.4.20
[root@securityonion ~]# rpm -qa |egrep "kernel-5|kernel-uek-5"
kernel-uek-5.15.0-103.114.4.el9uek.x86_64
kernel-5.14.0-284.25.1.0.1.el9_2.x86_64
kernel-uek-5.15.0-105.125.6.2.2.el9uek.x86_64
kernel-5.14.0-284.30.0.1.el9_2.x86_64
[root@securityonion ~]# df /boot
Filesystem     1K-blocks   Used Available Use% Mounted on
/dev/sda2         518816 417528    101288  81% /boot
```

After soup to 2.4.30:
```
[root@securityonion ~]# cat /etc/soversion 
2.4.30
[root@securityonion ~]# rpm -qa |egrep "kernel-5|kernel-uek-5"
kernel-uek-5.15.0-105.125.6.2.2.el9uek.x86_64
kernel-5.14.0-284.30.0.1.el9_2.x86_64
kernel-5.14.0-362.8.1.el9_3.x86_64
kernel-uek-5.15.0-200.131.27.el9uek.x86_64
[root@securityonion ~]# df -h /boot
Filesystem      Size  Used Avail Use% Mounted on
/dev/sda2       507M  410M   97M  81% /boot
```

**Reply by RagnarSTS** [reply]
No extra packages installed along the way; but you do seem to have confirmed that there are 4 kernels sitting there not 2.  
Your 2.4.5RC added a third kernel and 2.4.20 added a fourth.

Bumping from 2.4.20 > 2.4.30 did this?:
```
kernel-uek-5.15.0-103.114.4.el9uek.x86_64      >>  (removed)
kernel-uek-5.15.0-105.125.6.2.2.el9uek.x86_64  >>  kernel-uek-5.15.0-105.125.6.2.2.el9uek.x86_64 (nochange)
                                               >>  kernel-uek-5.15.0-200.131.27.el9uek.x86_64 (added)
kernel-5.14.0-284.25.1.0.1.el9_2.x86_64        >>  (removed)
kernel-5.14.0-284.30.0.1.el9_2.x86_64          >>  kernel-5.14.0-284.30.0.1.el9_2.x86_64 (nochange)
                                               >>  kernel-5.14.0-362.8.1.el9_3.x86_64 (added)
```
Guessing the process was add two updated kernels and only after successful completion remove the two older ones...might have been pinched there temporarily for some reason.

Even my lab setup it only failed out on one node, and didn't seem 100% full in the first place.  97M free on your example is indeed a little tight, but wouldn't normally cause a failure. 
Frequently as dedicated mount points /boot recommended minimum size is 1G and /boot/EFI would be 200M or so.   We're kind of splitting the difference in the middle at 500 and just need to be slightly more aware of limitations.

**Reply by dougburks** [reply]
> you do seem to have confirmed that there are 4 kernels sitting there not 2

The `installonly_limit=2` setting allows 2 versions of the standard RedHat kernel (the standard `kernel` package) and 2 versions of the Oracle kernel (the `kernel-uek` package).

> Bumping from 2.4.20 > 2.4.30 did this?:

Yes, it appears that way.

> Guessing the process was add two updated kernels and only after successful completion remove the two older ones...might have been pinched there temporarily for some reason.
Even my lab setup it only failed out on one node, and didn't seem 100% full in the first place. 97M free on your example is indeed a little tight, but wouldn't normally cause a failure.
Frequently as dedicated mount points /boot recommended minimum size is 1G and /boot/EFI would be 200M or so. We're kind of splitting the difference in the middle at 500 and just need to be slightly more aware of limitations.

I repeated my test above but this time with the VM set to EFI instead of BIOS and it worked fine without issues again. Full data below and it's almost identical to the first test run.

After installing 2.4.4 RC1:
```
[root@securityonion ~]# cat /etc/soversion
2.4.4
[root@securityonion ~]# rpm -qa |egrep "kernel-5|kernel-uek-5"
kernel-5.14.0-284.18.1.el9_2.x86_64
kernel-uek-5.15.0-103.114.4.el9uek.x86_64
[root@securityonion ~]# df -h /boot
Filesystem      Size  Used Avail Use% Mounted on
/dev/sda2       507M  272M  236M  54% /boot
```

After soup to 2.4.5:
```
[root@securityonion ~]# cat /etc/soversion 
2.4.5
[root@securityonion ~]# rpm -qa |egrep "kernel-5|kernel-uek-5"
kernel-5.14.0-284.18.1.el9_2.x86_64
kernel-uek-5.15.0-103.114.4.el9uek.x86_64
kernel-5.14.0-284.25.1.el9_2.x86_64
[root@securityonion ~]# df -h /boot
Filesystem      Size  Used Avail Use% Mounted on
/dev/sda2       507M  322M  185M  64% /boot
```

After soup to 2.4.10:
```
[root@securityonion ~]# cat /etc/soversion 
2.4.10
[root@securityonion ~]# rpm -qa |egrep "kernel-5|kernel-uek-5"
kernel-uek-5.15.0-103.114.4.el9uek.x86_64
kernel-5.14.0-284.25.1.el9_2.x86_64
kernel-5.14.0-284.25.1.0.1.el9_2.x86_64
[root@securityonion ~]# df -h /boot
Filesystem      Size  Used Avail Use% Mounted on
/dev/sda2       507M  318M  189M  63% /boot
```

After soup to 2.4.20:
```
[root@securityonion ~]# cat /etc/soversion 
2.4.20
[root@securityonion ~]# rpm -qa |egrep "kernel-5|kernel-uek-5"
kernel-uek-5.15.0-103.114.4.el9uek.x86_64
kernel-5.14.0-284.25.1.0.1.el9_2.x86_64
kernel-uek-5.15.0-105.125.6.2.2.el9uek.x86_64
kernel-5.14.0-284.30.0.1.el9_2.x86_64
[root@securityonion ~]# df -h /boot
Filesystem      Size  Used Avail Use% Mounted on
/dev/sda2       507M  403M  105M  80% /boot
```

After soup to 2.4.30:
```
[root@securityonion ~]# cat /etc/soversion 
2.4.30
[root@securityonion ~]# rpm -qa |egrep "kernel-5|kernel-uek-5"
kernel-uek-5.15.0-105.125.6.2.2.el9uek.x86_64
kernel-5.14.0-284.30.0.1.el9_2.x86_64
kernel-5.14.0-362.8.1.el9_3.x86_64
kernel-uek-5.15.0-200.131.27.el9uek.x86_64
[root@securityonion ~]# df -h /boot
Filesystem      Size  Used Avail Use% Mounted on
/dev/sda2       507M  405M  103M  80% /boot
```

**Reply by RagnarSTS** [reply]
Computers are silly.

**Reply by dougburks** [reply]
> Computers are silly.

Yes, teaching sand to think was a mistake!

In any case, we'll continue to keep an eye on this issue and see if anybody else runs into it.

**Reply by Acewiza** [reply]
It's like kids - Wadda ya expect given 'em dangerous toys like heat and light.

**Comment by RagnarSTS** [comment]
Perhaps use /boot/efi mountpoint for the 500M partition and keep the rest of /boot under normal / mountpoint.

**Comment by Acewiza** [comment]
I like the way the script tells you exactly what's wrong.  Bigger /boot maybe better, if you like keeping stuff hanging around.  Probably extrapolate a whole major version of updates based on x#fixes/month for a couple years without much of a disk premium these days.

**Comment by Sazerak** [comment]
I just came across this thread as I ran into the same issue. Tried running soup which failed due to lack of space on /boot. I did the checks suggested above but it only confirmed that there was in fact a lack of space. I ended up running dnf remove on the packages that were to be removed by dnf update and that seems to have fixed things (update is still going as I'm writing this, but seems to have made it farther than last try). This was on a standalone ISO install and the install limit was at 2.

**Reply by dougburks** [reply]
@Sazerak Just to confirm, are you using our Security Onion ISO image?

What version did you originally install?

What versions have you upgraded to since then?

Have you installed any other software?

Have you added any other files or directories to /boot?

As I mentioned above, I've tested this multiple times but have not been able to duplicate it. Are you able to provide steps to duplicate?

**Reply by Sazerak** [reply]
Hi Doug, to be honest SO2 has taken a bit of a back seat for me recently between family and work obligations, so I can't recall much of when/how I moved from 2.3 to 2.4, but what I can tell you is that it was a fresh ISO install which I think I started at 2.4.10. I probably upgraded to 2.4.20 before upgrading again today, but I can't say for sure. Nothing else has been installed and I'm not aware of anything else in /boot that shouldn't be there (nothing stood out when I checked).

Seeing as the upgrade worked now that I ran dnf remove as mentioned above, I don't think I'd be able to reproduce the steps, but if there are any logs or whatnot you'd like me to post here that you think might help, please don't hesitate to ask. I'm always happy to help! Unfortunately I didn't take screenshots, but I was lacking 19M in /boot which made the install fail. When I check with sudo dnf update I got the same results. Once I removed the excess kernels, the update went smoothly.

**Reply by dougburks** [reply, marked as answer]
Starting in 2.4.40, our ISO image will set /boot to 1GB:
https://github.com/Security-Onion-Solutions/securityonion/issues/12002

**Comment by Acewiza** [comment]
I just whacked the boot stuff I didn't want any longer.  ;-)
